lib/email/receiver.rb

```diff
@@ -801,44 +801,62 @@ def embedded_email_raw
-
-      raw = try_to_encode(embedded.decoded, "UTF-8").presence || embedded.to_s
-      title = embedded.subject.presence || subject
+      case SiteSetting.forwarded_emails_behaviour
+      when "create_replies"
+        forwarded_email_create_replies(destination, user)
+      when "quote"
+        forwarded_email_quote_forwarded(destination, user)
+      else
+        false
+      end
+    end
 
+    def forwarded_email_create_topic(destination, user, raw, title, date = nil)
       case destination[:type]
       when :group
         group = destination[:obj]
-        embedded_user = find_or_create_user(email, display_name)
-        post = create_topic(user: embedded_user,
-                            raw: raw,
-                            title: title,
-                            archetype: Archetype.private_message,
-                            target_usernames: [user.username],
-                            target_group_names: [group.name],
-                            is_group_message: true,
-                            skip_validations: true,
-                            created_at: embedded.date)
+        topic_user = block_given? ? yield : user
+        create_topic(user: topic_user,
+                     raw: raw,
+                     title: title,
+                     archetype: Archetype.private_message,
+                     target_usernames: [user.username],
+                     target_group_names: [group.name],
+                     is_group_message: true,
+                     skip_validations: true,
+                     created_at: date)
 
       when :category
         category = destination[:obj]
 
         return false if user.staged? && !category.email_in_allow_strangers
         return false if !user.has_trust_level?(SiteSetting.email_in_min_trust)
 
-        embedded_user = find_or_create_user(email, display_name)
-        post = create_topic(user: embedded_user,
-                            raw: raw,
-                            title: title,
-                            category: category.id,
-                            skip_validations: embedded_user.staged?,
-                            created_at: embedded.date)
+        topic_user = block_given? ? yield : user
+        create_topic(user: topic_user,
+                     raw: raw,
+                     title: title,
+                     category: category.id,
+                     skip_validations: topic_user.staged?,
+                     created_at: date)
       else
-        return false
+        false
       end
+    end
+
+    def forwarded_email_create_replies(destination, user)
+      embedded = Mail.new(embedded_email_raw)
+      email, display_name = parse_from_field(embedded)
+
+      return false if email.blank? || !email["@"]
+
+      raw = try_to_encode(embedded.decoded, "UTF-8").presence || embedded.to_s
+      title = embedded.subject.presence || subject
+
+      post = forwarded_email_create_topic(destination, user, raw, title, embedded.date) do
```

Comment:
Is there a reason to pass in the param via a block vs simply adding a `topic_user: nil` param?

Reply:
I did this to emulate the current logic around when a staged user is created:

```
      when :category
        category = destination[:obj]

        return false if user.staged? && !category.email_in_allow_strangers
        return false if !user.has_trust_level?(SiteSetting.email_in_min_trust)
```

originally the next line was:

```
embedded_user = find_or_create_user(email, display_name)
```

now the next line is:
```
topic_user = block_given? ? yield : user
```

If we pass the user in with a param, then we create the user prior to checking those conditions, and so I might be unnecessarily creating a staged user. I did try just creating the user and passing it as a param, but it caused a spec to fail.

Reply:
I see ... do we need this level of flexibility here? Maybe just pass email and display_name in optionally? something about this block is not sitting right for me... I think we should probably change the other occurrence as well but this is out of scope for this pr. If we must must have this flexibility I think passing in a lambda is a bit clearer intent wise.

Reply:
> Maybe just pass email and display_name in optionally?

You know, I hadn't actually thought of this, and jumped straight for the complicated solution. Whoops! Will update the PR.

Reply:
Actually I opted for a lambda because passing in email and display_name creates hell like this:

```
topic_user = embedded_user ? find_or_create_user(embedded_user[:email], embedded_user[:name]) : user
```

Splitting things over multiple lines doesn't make it much clearer, far less than what we get with a lambda (imo):

```
topic_user = embedded_user&.call || user
```